File: CPP/No173.cc
/**
 * Created by Xiaozhong on 2020/8/18.
 * Copyright (c) 2020/8/18 Xiaozhong. All rights reserved.
 */

#include "iostream"
#include "vector"

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class BSTIterator {
private:
    vector<int> ans;

    void dfs(TreeNode *root) {
        if (!root) return;
        if (root->left) dfs(root->left);
        ans.push_back(root->val);
        if (root->right) dfs(root->right);
    }

    int index = 0;

public:
    BSTIterator(TreeNode *root) {
        dfs(root);
    }

    /** @return the next smallest number */
    int next() {
        if (index < ans.size()) {
            return ans[index++];
        }
        return -1;
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return index < ans.size();
    }
};

File: Java/src/No173.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */


import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

class BSTIterator {
    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    List<Integer> data;
    Iterator<Integer> iterator;

    public BSTIterator(TreeNode root) {
        data = new ArrayList<>();
        dfs(root);
        iterator = data.iterator();
    }

    private void dfs(TreeNode root) {
        if (root == null) return;
        dfs(root.left);
        data.add(root.val);
        dfs(root.right);
    }

    public int next() {
        return iterator.next();
    }

    public boolean hasNext() {
        return iterator.hasNext();
    }
}

public class No173 {

}
